File: Week_08-位运算-布隆过滤-排序/493.翻转对.cpp
/*
 * @lc app=leetcode.cn id=493 lang=cpp
 *
 * [493] 翻转对
 */

// @lc code=start
class Solution {
private:
    vector<int> tmp;
    int mergeSort(vector<int>& nums, int l, int r) {
        //退出当前层逻辑
        if (l >= r) return 0;
        //进入下一层
        int mid = (l+r) >> 1;
        int cnt = mergeSort(nums, l, mid) + mergeSort(nums, mid+1, r);
        //处理当前层逻辑：找反转对
        int j=mid+1;
        //如果nums[i] 产生了j对， 那么nums[i+1]产生了j+对；
        for (int i = l; i <= mid; i++) {
            while (j <= r && (double)nums[i] > 2*(double)nums[j])
                j++;
            cnt += j - (mid + 1);
        }
        //处理当前层逻辑：合并两个子序列
        int i=l, k = 0;
        j=mid+1;
        while (i <= mid && j <= r) tmp[k++] = (nums[i]< nums[j]) ? nums[i++] : nums[j++];
        while (i <= mid) tmp[k++] = nums[i++];  
        while (j <= r) tmp[k++] = nums[j++];   
        for(int i=0; i<r-l+1; i++) nums[i+l] = tmp[i];
        return cnt;
    }
public:
    int reversePairs(vector<int>& nums) {
        tmp.resize(nums.size(), 0);
        int cnt = 0;
        cnt = mergeSort(nums, 0, nums.size()-1);
        return cnt;
    }
};
// @lc code=end

File: Week_08-位运算-布隆过滤-排序/493.翻转对.py
#
# @lc app=leetcode.cn id=493 lang=python
#
# [493] 翻转对
# 参考官方题解，参照c++版本改写而成
# 思路：归并排序


# @lc code=start
class Solution(object):
    def reversePairsRecursive(self, nums, left, right):
        if left == right:
            return 0
        else:
            mid = (left + right) // 2
            n1 = self.reversePairsRecursive(nums, left, mid)
            n2 = self.reversePairsRecursive(nums, mid+1, right)
            ret = n1 + n2
        # 首先统计下标对的数量
        i, j = left, mid + 1
        while i <= mid:
            while j <= right and nums[i] > 2* nums[j]:
                j += 1
            ret += (j - mid -1)
            i += 1
        
        # 随后合并两个排序数组
        sorted = [0 for _ in range(right-left+1)]
        p1, p2 = left, mid + 1
        p = 0
        while p1<=mid or p2<=right:
            if p1 > mid:
                sorted[p] = nums[p2]
                p += 1
                p2 += 1
            elif p2 > right:
                sorted[p] = nums[p1]
                p += 1
                p1 += 1
            else:
                if nums[p1] < nums[p2]:
                    sorted[p] = nums[p1]
                    p += 1
                    p1 += 1
                else:
                    sorted[p] = nums[p2]
                    p += 1
                    p2 += 1
        for i in range(len(sorted)):
            nums[left+i] = sorted[i]
        return ret

    def reversePairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        return self.reversePairsRecursive(nums, 0, len(nums)-1)

# @lc code=end

